great_expectations/rule_based_profiler/parameter_builder/regex_pattern_string_parameter_builder.py

```diff
@@ -0,0 +1,193 @@
+    build_parameter_container,
+)
+from great_expectations.rule_based_profiler.util import (
+    get_parameter_value_and_validate_return_type,
+)
+from great_expectations.validator.validator import Validator
+
+logger = logging.getLogger(__name__)
+
+
+class RegexPatternStringParameterBuilder(ParameterBuilder):
+    """
+    Detects the domain REGEX from a set of candidate REGEX strings by computing the
+    column_values.match_regex_format.unexpected_count metric for each candidate format and returning the format that
+    has the lowest unexpected_count ratio.
+    """
+
+    # list of candidate strings that are most commonly used
+    # source: https://regexland.com/most-common-regular-expressions/
+    CANDIDATE_REGEX: Set[str] = {
+        r"/\d+/",  # whole number with 1 or more digits ExpectValuesToBeNumeric? (.. youw oudl want to emit that expectation)?
+        r"/-?\d+/",  # negative whole numbers
+        r"/-?\d+(\.\d*)?/",  # decimal numbers with . (period) separator
+        r"/[A-Za-z0-9\.,;:!?()\"'%\-]+/",  # general text
+        r"^ +/",  # leading space
+        r" +/$",  # trailing space
+        r"/https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#()?&//=]*)/",  #  Matching URL (including http(s) protocol)
+        r"/<\/?(?:p|a|b|img)(?: \/)?>/",  # HTML tags
+        r"/(?:25[0-5]|2[0-4]\d|[01]\d{2}|\d{1,2})(?:.(?:25[0-5]|2[0-4]\d|[01]\d{2}|\d{1,2})){3}/",  # IPv4 IP address
+        r"/(?:[A-Fa-f0-9]){0,4}(?: ?:? ?(?:[A-Fa-f0-9]){0,4}){0,7}/",  # IPv6 IP address,
+    }
+
+    def __init__(
+        self,
+        name: str,
+        metric_domain_kwargs: Optional[Union[str, dict]] = None,
+        metric_value_kwargs: Optional[Union[str, dict]] = None,
+        threshold: float = 1.0,
+        candidate_regexes: Optional[Union[str, dict]] = None,
+        data_context: Optional["DataContext"] = None,
+        batch_request: Optional[Union[BatchRequest, RuntimeBatchRequest, dict]] = None,
+    ):
+        """
+        Configure this RegexPatternStringParameterBuilder
+        Args:
+            name: the name of this parameter -- this is user-specified parameter name (from configuration);
+            it is not the fully-qualified parameter name; a fully-qualified parameter name must start with "$parameter."
+            and may contain one or more subsequent parts (e.g., "$parameter.<my_param_from_config>.<metric_name>").
+            threshold: the ratio of values that must match a format string for it to be accepted
+            candidate_regexes: a list of candidate regex strings that will REPLACE the default
+            data_context: DataContext
+            batch_request: specified in ParameterBuilder configuration to get Batch objects for parameter computation.
+        """
+        super().__init__(
+            name=name,
+            data_context=data_context,
+            batch_request=batch_request,
+        )
+
+        self._metric_domain_kwargs = metric_domain_kwargs
```

Comment:
@Shinnnyshinshin This needs to change somewhat -- if you could please see the flavor in the latest `[parameter_builder/simple_date_format_string_parameter_builder.py]` and/or our numeric/metric parameter builders.  Specifically:
* The properties have their "raw" type union with `str`.
* Their actual values are obtained "in the flow of the logic" with the correct type using the parameter getter, for example:
```
        # Obtain reduce_scalar_metric from "rule state" (i.e., variables and parameters); from instance variable otherwise.
        reduce_scalar_metric: bool = get_parameter_value_and_validate_return_type(
            domain=domain,
            parameter_reference=self.reduce_scalar_metric,
            expected_return_type=bool,
            variables=variables,
            parameters=parameters,
        )
```
Please notice that the type of the variable changes from Union[its_type, str] to its specific to just `its_type` (and the variable is of method, not of instance).  This gives us the most flexibility in retrieving properties from `$variable` and `$parameter` storages.

Further, please create property accessors for all private variables -- this is critical for compliance with the `DictDot` class hierarchy (and thus gaining the benefits they bring).  I am happy to pair on this, if needed.  Thanks!

Reply:
Done. Thank you for the thorough explanation @alexsherstinsky